orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc

```diff
@@ -0,0 +1,550 @@
+      auto axis = static_cast<int64_t>(cur->GetAttributes().at("axis").i());
+      axis = axis < 0 ? axis + cur->InputDefs()[0]->Shape()->dim_size() : axis;
+      if (axis < 2) {
+        LOG_DEBUG_INFO(logger, "PaddingElimination::axis of Normalization: " + cur->Name() + " is " +
+                                   std::to_string(axis) + ", which blocks merging leading two dims.");
+        candidate_output.insert(cur);
+      } else {
+        candidate_edges.insert(cur->MutableOutputDefs()[0]);
+        PushAllOutputNode(graph, to_visit, cur, visited);
+      }
+    } else if (graph_utils::IsSupportedOptypeVersionAndDomain(*cur, "Dropout", {12, 13})) {
+      ORT_ENFORCE(candidate_edges.find(cur->MutableInputDefs()[0]) != candidate_edges.end());
+      candidate_edges.insert(cur->MutableOutputDefs()[0]);
+      candidate_edges.insert(cur->MutableOutputDefs()[1]);
+      PushAllOutputNode(graph, to_visit, cur, visited);
+    } else if (graph_utils::IsSupportedOptypeVersionAndDomain(*cur, "Cast", {9, 13}) ||
+               graph_utils::IsSupportedOptypeVersionAndDomain(*cur, "BiasGelu", {1}, kMSDomain)) {
+      ORT_ENFORCE(candidate_edges.find(cur->MutableInputDefs()[0]) != candidate_edges.end());
+      candidate_edges.insert(cur->MutableOutputDefs()[0]);
+      PushAllOutputNode(graph, to_visit, cur, visited);
+    } else if (graph_utils::IsSupportedOptypeVersionAndDomain(*cur, "MatMul", {1, 9, 13})) {
+      ORT_ENFORCE(candidate_edges.find(cur->MutableInputDefs()[0]) != candidate_edges.end() ||
+                  candidate_edges.find(cur->MutableInputDefs()[1]) != candidate_edges.end());
+      if (candidate_edges.find(cur->MutableInputDefs()[0]) != candidate_edges.end()) {
+        // If shape of [batch_size, seqlen, ...] is propagated from the first argument of MatMul.
+        // The dim size of the first argument must larger than 2 to propagete the first two dims to the output.
+        // Or else the first two dims of the output will not be [batch_size, seqlen] and this MatMul will be added
+        // to candidate_output as the output of the subgraph.
+        if (cur->InputDefs()[0]->Shape()->dim_size() > 2) {
+          candidate_edges.insert(cur->MutableOutputDefs()[0]);
+          PushAllOutputNode(graph, to_visit, cur, visited);
+        } else {
+          LOG_DEBUG_INFO(logger,
+                         "PaddingElimination::dim size of left input of matmul smaller than 3 and \
+                            this matmul would be output of subgraph.");
+          candidate_output.insert(cur);
+          continue;
+        }
+      } else {
+        LOG_DEBUG_INFO(logger, "PaddingElimination::right edge of matmul would not included.");
+        candidate_output.insert(cur);
+        continue;
+      }
+    } else if (graph_utils::IsSupportedOptypeVersionAndDomain(*cur, "PythonOp", {1}, kMSDomain)) {
+      ORT_ENFORCE(candidate_edges.find(cur->MutableInputDefs()[0]) != candidate_edges.end());
```

Comment:
I ton't think this will alway hold "ORT_ENFORCE(candidate_edges.find(cur->MutableInputDefs()[0]) != candidate_edges.end());", can we tolerant if this condition not meet, continue, instead of fail the program.